gazenot/src/lib.rs

```diff
@@ -0,0 +1,653 @@
+            let handle = self.clone();
+            let releases = releases
+                .iter()
+                .map(|r| AnnounceReleaseKey {
+                    package: r.package.clone(),
+                    tag: r.tag.clone(),
+                })
+                .collect();
+            let announcement = announcement.clone();
+            queries.push(tokio::spawn(async move {
+                handle
+                    .create_announcement(url, releases, announcement)
+                    .await
+            }));
+        }
+
+        // Then join on them all
+        for query in queries {
+            query.await??;
+        }
+        Ok(())
+    }
+
+    async fn create_announcement(
+        &self,
+        url: Url,
+        releases: Vec<AnnounceReleaseKey>,
+        announcement: AnnouncementKey,
+    ) -> Result<()> {
+        let request = AnnounceReleaseRequest {
+            releases,
+            body: announcement.body,
+        };
+        let response = self
+            .client
+            .post(url.clone())
+            .header("AUTHORIZATION", &self.api_key)
+            .json(&request)
+            .send()
+            .await
+            .and_then(|res| res.error_for_status())?;
+        let text = axoasset::SourceFile::new(url.as_str(), response.text().await?);
+        let parsed: BasicResponse = text.deserialize_json()?;
+        parsed.result()
+    }
+
+    pub fn create_artifact_set_url(&self, package: &PackageKey) -> Result<Url> {
+        // POST /api/:owner/:package/artifacts
+        let server = &self.api_server;
+        let PackageKey { owner, package } = package;
+        let url = Url::from_str(&format!("https://{server}/api/{owner}/{package}/releases"))?;
+        Ok(url)
+    }
+
+    pub fn download_artifact_set_url(&self, set: &ArtifactSet, filename: &str) -> Result<Url> {
+        // GET :owner.:hosting_server/:package/:public_id/
+        let base = set.set_download_url.clone().unwrap_or_else(|| {
+            let server = &self.hosting_server;
+            let ArtifactSet {
+                owner,
+                package,
+                public_id,
+                ..
+            } = set;
+            format!("https://{owner}.{server}/{package}/{public_id}")
+        });
+        let url = Url::from_str(&format!("{base}/{filename}"))?;
+        Ok(url)
+    }
+
+    pub fn upload_artifact_set_url(&self, set: &ArtifactSet, filename: &str) -> Result<Url> {
+        // POST /api/:owner/:package/artifacts/:public_id/upload/:filename
+        let base = set.upload_url.clone().unwrap_or_else(|| {
+            let server = &self.api_server;
+            let ArtifactSet {
+                owner,
+                package,
+                public_id,
+                ..
+            } = set;
+            format!("https://{server}/api/{owner}/{package}/artifacts/{public_id}/upload")
+        });
+        let url = Url::from_str(&format!("{base}/{filename}"))?;
```

Comment:
Can we use `join` on a parsed `Url` for this? https://docs.rs/reqwest/0.11.22/reqwest/struct.Url.html#method.join In general I wonder if we can get away with less string formatting/interpolation here

Reply:
So all the URL path APIs are regrettably DEEPLY CURSED and basically unusable as a result. Unlike the std::path APIs which basically say "oh ok, if you join/push a component, we'll assume the lhs was a dir" the URL crate will only do that if the input ends in a trailing slash. Otherwise it will clobber the last component, on the assumption that it was a file.

This is why oranda has been plagued with bugs that are like "oh shit we forgot a trailing slash and everything broke".

So at that point you're finding yourself needing to be super pedantic about the format of your input anyway, and it's easier/clearer to just use string formatting 🙃